contracts/strategies/BenqiStrategyV1.sol

```diff
@@ -0,0 +1,357 @@
+        WAVAX.deposit{value: address(this).balance}();
+        uint qiBalance = rewardToken0.balanceOf(address(this));
+        DexLibrary.swap(qiBalance, address(rewardToken0), address(rewardToken1), swapPairToken0);
+        amount = rewardToken1.balanceOf(address(this));
+
+        uint devFee = amount.mul(DEV_FEE_BIPS).div(BIPS_DIVISOR);
+        if (devFee > 0) {
+            _safeTransfer(address(rewardToken), devAddr, devFee);
+        }
+
+        uint adminFee = amount.mul(ADMIN_FEE_BIPS).div(BIPS_DIVISOR);
+        if (adminFee > 0) {
+            _safeTransfer(address(rewardToken), owner(), adminFee);
+        }
+
+        uint reinvestFee = amount.mul(REINVEST_REWARD_BIPS).div(BIPS_DIVISOR);
+        if (reinvestFee > 0) {
+            _safeTransfer(address(rewardToken), msg.sender, reinvestFee);
+        }
+        uint depositTokenAmount = DexLibrary.swap(
+            amount.sub(devFee).sub(adminFee).sub(reinvestFee),
+            address(rewardToken1),
+            address(depositToken),
+            swapPairToken1
+        );
+
+        _stakeDepositTokens(depositTokenAmount);
+        totalDeposits = totalDeposits.add(depositTokenAmount);
+
+        emit Reinvest(totalDeposits, totalSupply);
+    }
+
+    function _rollupDebt(uint principal, uint borrowed) internal {
+        (uint borrowLimit, uint borrowBips) = _getBorrowLimit();
+        uint supplied = principal;
+        uint lendTarget = principal.sub(borrowed).mul(leverageLevel).div(leverageBips);
+        uint totalBorrowed = borrowed;
+        while(supplied < lendTarget) {
+            uint toBorrowAmount = _getBorrowable(supplied, totalBorrowed, borrowLimit, borrowBips);
+            if (supplied.add(toBorrowAmount) > lendTarget) {
+                toBorrowAmount = lendTarget - supplied;
+            }
+            // safeguard needed because we can't mint below a certain threshold
+            if (toBorrowAmount < minMinting) {
+                toBorrowAmount = minMinting;
+            }
+            require(tokenDelegator.borrow(toBorrowAmount) == 0, "BenqiStrategyV1::borrowing failed");
+            require(tokenDelegator.mint(toBorrowAmount) == 0, "BenqiStrategyV1::lending failed");
+            supplied = tokenDelegator.balanceOfUnderlying(address(this));
+            totalBorrowed = tokenDelegator.borrowBalanceCurrent(address(this));
+        }
+    }
+
+    function _getRedeemable(uint balance, uint borrowed, uint borrowLimit, uint bips) internal pure returns (uint256) {
+        return balance.sub(borrowed.mul(bips).div(borrowLimit));
+    }
+
+    function _getBorrowable(uint balance, uint borrowed, uint borrowLimit, uint bips) internal pure returns (uint256) {
+        return balance.mul(borrowLimit).div(bips).sub(borrowed);
+    }
+
+    function _getBorrowLimit() internal view returns (uint, uint) {
+        (, uint borrowLimit) = rewardController.markets(address(tokenDelegator));
+        return (borrowLimit, 1e18);
```

Comment:
is `1e18` determined by the underlying token decimals?

Reply:
no, according to compound doc https://compound.finance/docs/comptroller#collateral-factor it's hardcoded because of their mantissa